SJRK-289: Add server-side auto-save to include files

BlueSlug opened the pull request:
This work implements server-side auto-saving of the story during its creation. This required totally reworking the way stories are created and published in the first place: rather than saving to database at publish time, the story is saved when the Edit page is first loaded, then when the user hits "publish", it switches a flag to indicate the story is ready for public consumption.

This work also encapsulates the following issues to varying degrees:
- [SJRK-185](https://issues.fluidproject.org/browse/SJRK-185) is dealt with, using UUID v4 now
- [SJRK-175](https://issues.fluidproject.org/browse/SJRK-175) is partly done, as the `videoBlockEditor` grade is now a `mediaBlockEditor`. Much remains to be done on this Jira, to be handled on another occasion

BlueSlug committed:
SJRK-289: restricted Browse page db query to show published stories only

BlueSlug committed:
SJRK-289: set version number to 1.0.0
This update is a breaking change, since we'll need to modify 
previously-existing stories to continue to work after this update.

BlueSlug committed:
SJRK-289: added timestamp fields back into story model, specified format

BlueSlug committed:
SJRK-185: updated UUID generation to use v4 (random basis, not datetime)

BlueSlug committed:
SJRK-289: restrict View response to load only published stories

BlueSlug committed:
SJRK-289: added basic testing for Viewing an unpublished story

BlueSlug committed:
Merge remote-tracking branch 'upstream/master' into SJRK-289

BlueSlug committed:
SJRK-289: stories now save without files, tests updated to match
Includes the new test function `mockFileUpload` which may be removed 
once individual files can be posted

BlueSlug committed:
SJRK-289: `submitStory` fixed after server changes

BlueSlug committed:
SJRK-289: WIP - setting up story initialization and refactoring autosave This includes drawing a distinction in code between saving the story to the server and publishing, and it sets up a number of invokers and events in order to make these separate tasks more configurable.

This commit leaves the Edit Page tests in a broken state.

BlueSlug committed:
SJRK-289: WIP - adjusted publishing event configuration and fixed tests

BlueSlug committed:
SJRK-289: solved bug that disappeared published stories under new IDs

BlueSlug committed:
SJRK-289: some more Edit page refactoring and fixing tests

BlueSlug committed:
SJRK-289: saveStoryWithBinariesHandler -> saveStoryHandler, updated docs

BlueSlug committed:
SJRK-289: WIP added saveStoryFile handlers based on previous handlers

BlueSlug committed:
SJRK-289: WIP fixed references to saveStoryFile middleware, added tests
The server tests with storage are currently broken

BlueSlug committed:
SJRK-289: fixed saveStoryFile middleware & requestHandler issues, tests

BlueSlug committed:
SJRK-289: removed broken tests for buildBinaryRenameMap

BlueSlug committed:
SJRK-289: expanded server tests for story with image block, publishing

BlueSlug committed:
SJRK-289: removed client-side image orientation code, dependencies
Manual image block editor tests remain unchanged as we want to ensure 
the server-side orientation still works, once that's all wired up.

BlueSlug committed:
SJRK-289: WIP refactoring uploader grade to handle file uploading

BlueSlug committed:
Merge remote-tracking branch 'upstream/master' into SJRK-289

BlueSlug committed:
SJRK-289: wired up file uploading and preview update for image blocks

BlueSlug committed:
SJRK-289: removed clearMediaBlockUrls & associated tests

BlueSlug committed:
SJRK-289: updated handleGetStory to leave block file URLs as-is

BlueSlug committed:
SJRK-175: videoBlockEditor is now a mediaBlockEditor for more DRYness

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -105,36 +105,28 @@ fluid.defaults("sjrk.storyTelling.server.getStoryHandler", {
+        } else {
+            request.events.onError.fire({
+                isError: true,
+                message: "An error occurred while retrieving the requested story"
```
It would be helpful to provide a clearer error message such as "Unauthorized: cannot access an unpublished story". If you'd rather not to expose the stories status to the end user, perhaps just return the unauthorized error but track the actual reason of being unauthorized in the server log.

BlueSlug replied:
I've been going back and forth on this one, not sure whether I should be worried about enumeration attacks. Since we're using v4 UUID's, it's not impossible to guess ID's, but it's going to be very difficult and not a very practical vector, IMO.

If I'd like to avoid giving away that a story ID was correct but that it's unpublished, then it would be better to match this error message to the one below it. I'll make them both the same and send the actual error message on the server log, in each case.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -199,20 +166,24 @@ sjrk.storyTelling.server.handleSaveStoryWithBinaries = function (request, dataSo
-sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenameMap, storyModel, successEvent, failureEvent) {
-    var id = uuidv1();
+sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, storyModel, successEvent, failureEvent) {
+    var id = "";
```
line 178-184 can be simplified to: 
```
var id = storyModel && storyModel.id ? storyModel.id : uuidv4();
```

BlueSlug replied:
Good catch! I'd left it that way to easily set breakpoints, but that's no longer necessary

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -222,45 +193,86 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
         }
-    });
+    } else {
+        // delete the file?
```
Can this comment be removed as the file hasn't been saved?

BlueSlug replied:
The file is actually saved before this requestHandler is ever called, so it does in fact persist in the server file system even if authoring is disabled.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -222,45 +193,86 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
+
+/**
+ * Saves a single file to the server filesystem and responds to the HTTP request.
+ * If authoring is not enabled, the request raises an error.
```
In this comment, please cover all cases that an error will be raised.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -222,45 +193,86 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
+// Kettle request handler for saving a single file associated with a pre-existing story
+fluid.defaults("sjrk.storyTelling.server.saveStoryFileHandler", {
+    gradeNames: "kettle.request.http",
+    // a DataSource to save a single story along with any files it has
```
This comment is inaccurate. Probably remove it as the comment at line 196 is more accurate.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -222,45 +193,86 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
+ * If authoring is not enabled, the request raises an error.
  *
- * @return {Object.<String, String>} - a key-value collection linking uploaded filenames to server paths
+ * @param {Object} request - a Kettle request with data and files associated with a single story
```
This request only contains one single file. Please update. Thanks.

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -149,6 +142,51 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
                 funcName: "sjrk.storyTelling.base.page.storyEdit.clearAutosave",
                 args: ["{that}.options.pageSetup.storyAutosaveKey"]
+            },
+            initializeStory: {
```
Please add code comments to explain how this and following functions are used at different stages of the auto save and publish.

BlueSlug committed:
SJRK-289: audio & video blocks now upload files and update properly

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -222,45 +193,86 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
-                    sjrk.storyTelling.server.setMediaBlockUrl(block, null);
+sjrk.storyTelling.server.handleSaveStoryFile = function (request, dataSource, authoringEnabled) {
+    if (authoringEnabled) {
+        try {
```
This `try/catch` is not quite necessary as the reject callback of `getStoryPromise` will catch most failure. Once `id` at line 225 is checked, the `try/catch` can be removed and this deeply nested block will have one level less. To do so, line 222-226 can be refactored to:

```
var id = request.req.params.id;
if (authoringEnabled && id) {
    var getStoryPromise = ...
} else {
    // report error
};
```

BlueSlug replied:
I've actually removed this part and refactored the function entirely, since the server was totally ignoring the try/catch and crashing anyway

cindyli commented on src/server/middleware/saveStoryFile.js:
```diff
@@ -55,8 +48,8 @@ fluid.defaults("sjrk.storyTelling.server.middleware.saveStoryWithBinaries", {
-sjrk.storyTelling.server.middleware.saveStoryWithBinaries.filenameResolver = function (req, file, cb) {
-    var id = uuidv1();
+sjrk.storyTelling.server.middleware.saveStoryFile.filenameResolver = function (req, file, cb) {
+    var id = uuidv4();
```
A problem with generating a new id for every uploaded file is, if a user operates on a same media block by changing the file a few times, all these files will be kept on the server even though the last one is the one that should be kept. Those previous uploads won't be able to be removed even by deleting the story because the story only has the url of the last upload for that media block.

One solution is to generate this id and compose the file name at the client side. The server removes files with the same id before saving the new file. The removal is needed because the file extension can be different.

BlueSlug replied:
Indeed, I was thinking of passing in model detail containing the current filename in order to remove that from the server before operating on the new file. Since this is multi-part form data, it should be feasible.

I'd also considered creating the ID client-side, but decided to keep all ID generation on the server.

BlueSlug committed:
SJRK-289: Added progress indicator & related events to imageBlockEditor

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -274,50 +312,70 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+            // store the ID on the story model for later use
+            story.applier.change(storyIdPath, successResponse.id, null, sourceName);
+        }, function (jqXHR, textStatus, errorThrown) {
+            fluid.log(fluid.logLevel.WARN, "Something went wrong");
```
Be more specific with "Something went wrong". Replace it with a clearer error such as "Error at saving a new story to server."

BlueSlug committed:
SJRK-289: reworked handleGetStory error reporting to prevent enum attack
The true error detail will be preserved on the server log

cindyli commented on src/ui/uploader.js:
```diff
@@ -129,11 +155,11 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
-            that.applier.change("fileObjectURL", URL.createObjectURL(currentFile));
+            URL.revokeObjectURL(that.model.fileObjectUrl);
+            that.applier.change("fileObjectUrl", URL.createObjectURL(currentFile));
             that.applier.change("fileDetails", fileDetails);
```
Is `fileDetails` saved for the testing purpose? It doesn't seem to be used in the implementation code.

BlueSlug replied:
It was formerly used quite heavily in the `binaryRenameMap` collections that were passed around in order to associate the various story blocks with the files that were uploaded en masse on story publication. It's entirely possible that it isn't needed any more.

It seems it is only used in `testServerWithStorage.js`, now. I'll try to remove it and see if anything breaks :)

cindyli commented on tests/server/testServerWithStorage.js:
```diff
@@ -611,6 +761,31 @@ sjrk.storyTelling.server.testServerWithStorageDefs.verifyStoryPersistence = func
+    var parsedData = JSON.parse(data);
+
+    if (authoringEnabled) {
+        jqUnit.assertEquals("Story request produced an error as expected", true, parsedData.isError);
```
Please verify error message, at least partial message, to make sure the rejection is due to the unpublished state.

cindyli commented on tests/server/testServerWithStorage.js:
```diff
@@ -611,6 +761,31 @@ sjrk.storyTelling.server.testServerWithStorageDefs.verifyStoryPersistence = func
+            completionEvent.fire(parsedData);
+        }
+    } else {
+        jqUnit.assert("Saved story data does not exist");
```
Shall here verify an error message if `authoringEnabled === false` rather than a fake `assert`?

cindyli commented on tests/server/testServerWithStorage.js:
```diff
@@ -327,23 +376,49 @@ sjrk.storyTelling.server.testServerWithStorageDefs = [{
     },
     // Story with an image
     {
+        // initializes the story in its unpublished form
```
Line 373-616 is a very long sequence. It should be broken down to a few sequences, each testing a meaning feature.

BlueSlug replied:
I agree that it's a little bit spaghetti, with each test feeding into the next. Breaking it into sequences would also allow for more variety in the test cases.

Doing this will take a fair amount of time, I believe, so we may want to revisit the release date in that case.

cindyli commented on tests/server/testServerWithStorage.js:
```diff
@@ -562,39 +716,35 @@ sjrk.storyTelling.server.testServerWithStorageDefs.getSavedStory = function (sto
-
         if (completionEvent && fileOptions) {
             completionEvent.fire(parsedData.content[0][fileOptions.urlProp]);
         }
```
Shall the else block at line 751 to verify an error response when `authoringEnabled === false`?

cindyli commented on tests/ui/js/base-page-storyEditTests.js:
```diff
@@ -131,16 +138,21 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
             }
         ],
         "tags": ["cute", "cats"],
+        "timestampCreated": "",
```
Don't forget to update `timestampCreated`, `timestampLastModified` and `timestampPublished` to proper values at proper stages.

BlueSlug replied:
Good catch, I'd forgotten about these three :) Thanks!

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -109,30 +115,17 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
-                    "{that}.options.pageSetup.storyAutosaveKey"
-                ]
-            },
+            "onStorySaveToServerRequested.saveStoryToServer": "{that}.saveStoryToServer",
```
`onStorySaveToServerRequested` to save the auto save to the server hasn't been called in this pull request. I guess it'll be in the later commits.

BlueSlug replied:
I added these three events together, but my plan for each of them has changed a little. I don't anticipate using `onStorySaveToServerRequested` in my current work for this Jira. Should I remove it, in that case?

cindyli replied:
I think it's a good idea to remove it and only add when it'll be used. Thanks.

BlueSlug committed:
SJRK-289: Addressed pull request review
- updated some comments
- tidied up some logic that was left long-form for debugging purposes
- renamed `loadStoryFromAutosave` to `loadStoryContent` to be more 
descriptive of what it actually does
- improved error reporting
- updated server storage tests

BlueSlug committed:
SJRK-289: setting timestamps for creation and publishing, fixed tests

BlueSlug committed:
SJRK-289: removed onStorySaveToServerRequested as it was not in use

BlueSlug committed:
SJRK-289: updated filenameResolver to include storyId, fixed error msg

BlueSlug committed:
SJRK-289: WIP added removePreviousStoryFile middleware, began wiring it

BlueSlug committed:
SJRK-289: implemented timestampLastModified as model listener on story

BlueSlug committed:
SJRK-289: added progress indicator to video and audio blocks. WET code..

BlueSlug committed:
SJRK-289: fixed server tests (added missing reference to middleware)

BlueSlug committed:
SJRK-289: got removePreviousStoryFile middleware working properly

BlueSlug committed:
SJRK-289: removed now-unused fileDetails references, updated server test

BlueSlug committed:
SJRK-289: DRYed up the uploader state management a little bit. Still WET

BlueSlug committed:
SJRK-175/289: heavy refactoring of image and media blocks, see detail:
- server and browser tests are passing, but blocks aren't working quite 
right (yet)
- `imageUrl` is now `mediaUrl`
- all associated logic to parse block type is removed, including tests 
for the same
- new block editor `withFileUploader` grade to manage the file uploader 
and progress indicator
- references to `mediaPlayer` now called `mediaPreview`

BlueSlug committed:
SJRK-289: solved model relay bug with file uploader visibility states

BlueSlug committed:
Merge remote-tracking branch 'upstream/master' into SJRK-289

jobara commented on src/server/requestHandlers.js:
```diff
@@ -199,20 +172,18 @@ sjrk.storyTelling.server.handleSaveStoryWithBinaries = function (request, dataSo
-sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenameMap, storyModel, successEvent, failureEvent) {
-    var id = uuidv1();
+sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, storyModel, successEvent, failureEvent) {
+    var id = storyModel && storyModel.id ? storyModel.id : uuidv4();
```
You could also write this as `fluid.get(storyModel, "id") || uuidv4()`

jobara commented on src/server/requestHandlers.js:
```diff
@@ -222,45 +193,85 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
+                error: err
+            });
+        });
+    } else {
+        request.events.onError.fire({
+            isError: true,
+            message: "Saving is currently disabled."
+        });
+    }
```
It might be easier to read this if you move it to the top and just have a simple if condition that returns. 

```javascript
if (!authoringEnabled) { 
    request.events.onError.fire({
        isError: true,
        message: "Saving is currently disabled."
    });
    return;  
}
```

jobara commented on src/server/requestHandlers.js:
```diff
@@ -222,45 +193,85 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
-                if (mediaFile) {
-                    sjrk.storyTelling.server.setMediaBlockUrl(block, mediaFile.filename);
-                    binaryRenameMap[mediaFile.originalname] = mediaFile.filename;
+sjrk.storyTelling.server.handleSaveStoryFile = function (request, dataSource, authoringEnabled) {
```
I'm not sure I'm following the workflow of this function. It says it handles saving, but I'm not exactly sure where that happens. It seems that it is already saved by the time this handler executes and that this function will attempt to rotate the image as needed and also verifies the save. Is that correct?

BlueSlug committed:
Merge remote-tracking branch 'upstream/master' into SJRK-289
			
The manual tests for audioBlockEditor, imageBlockEditor and 
videoBlockEditor are no longer valid, as the work on SJRK-289 requires 
that the CouchDB server is running, while the work on SJRK-122 has 
isolated the manual tests from the Kettle server, thus cutting them off 
from CouchDB. I have chosen to leave the files in place until an 
appropriate solution is found.

BlueSlug committed:
SJRK-289: adding missing JSDoc parameters, clarifying variable name

BlueSlug committed:
SJRK-289: improve readability of handleSaveStoryFile function

BlueSlug committed:
SJRK-289: fixed bug in uploader's valueMapper model relay specification

BlueSlug committed:
Merge remote-tracking branch 'upstream/master' into SJRK-289

BlueSlug committed:
SJRK-289: add more detailed upload call err messages, fix a filename bug

Also adds a rudimentary timeout setup. Should be parameterized and 
configurable in the server config file.

cindyli commented on src/server/middleware/removePreviousStoryFile.js:
```diff
@@ -0,0 +1,49 @@
+
+        fluid.log(fluid.logLevel.WARN, "Deleting file: " + fileToDelete);
+
+        if (fs.existsSync(fileToDelete)) {
```
In the case the previous file doesn't exist, might be helpful to log this message.

When the file does get removed successfully, log the success too.

jobara commented on src/server/middleware/removePreviousStoryFile.js:
```diff
@@ -0,0 +1,49 @@
+var sjrk = fluid.registerNamespace("sjrk");
+
+// Middleware to remove the previous file associated with a story block
+fluid.defaults("sjrk.storyTelling.server.middleware.removePreviousStoryFile", {
```
Consider renaming, as it appears that this can remove any file, not just the previous one.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -111,50 +111,43 @@ fluid.defaults("sjrk.storyTelling.server.getStoryHandler", {
+            });
+        }
     }, function (error) {
         var errorAsJSON = JSON.stringify(error);
```
errorAsJSON -> errorInStr?

jobara replied:
This seems to be misnamed as it's a string value, not JSON.

jobara commented on src/server/requestHandlers.js:
```diff
@@ -168,29 +161,9 @@ fluid.defaults("sjrk.storyTelling.server.saveStoryWithBinariesHandler", {
  * @param {Boolean} authoringEnabled - a server-level flag to indicate whether authoring is enabled
  */
-sjrk.storyTelling.server.handleSaveStoryWithBinaries = function (request, dataSource, authoringEnabled) {
+sjrk.storyTelling.server.handleSaveStory = function (request, dataSource, authoringEnabled) {
```
Rather than having flags at the handler level for `authoringEnabled` we should modify the configuration, probably using context awareness, to not include configuration related to editing, when authoring is disabled.

BlueSlug replied:
As we discussed, this is out of scope for this Jira. I have filed [SJRK-415](https://issues.fluidproject.org/browse/SJRK-415) to work on this improvement at a later date.

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -109,58 +114,98 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                ]
+            },
+            // Saves the current story to the server
+            saveStoryToServer: {
```
With the use of the `saveNewStoryToServer()` name, `saveStoryToServer()` gives an impression of saving an existing story to the server. But this function is to make an ajax call to save any story. It probably make sense to clarify the name such as `postStoryToServer`?

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -384,55 +452,72 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+    sjrk.storyTelling.base.page.storyEdit.publishStory = function (storyEditPage, viewPageUrl, story, successCallback, errorEvent) {
+        // set the publish timestamp to now and the flag to "true"
+        story.applier.change("timestampPublished", new Date().toISOString());
+        story.applier.change("published", true);
```
Consider to optimize the change request to:

```suggestion
story.applier.change("", {
    "timestampPublished": new Date().toISOString(),
    "published": true
});
```

The benefit is when there's a model listener listening to the root of the model, this listener won't be fired twice.

cindyli commented on src/ui/blockUi-editor-audioBlockEditor.js:
```diff
@@ -16,14 +16,24 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+        selectors: {
+            previewPlaceholder: ".sjrkc-st-block-audio-preview-placeholder"
+        },
+        modelListeners: {
```
Could you please add a code comment to explain why this model listener is required? Thanks.

BlueSlug replied:
I tried to determine why this was only in place for the audio block editor, and pored over my notes and change logs from when I added this model listener (back on May 20), and it seems that at the time the audio block was the only type that required its own explicit model listener to hide and show the preview placeholder image (the little microphone icon) along with the preview element itself (in this case, the `audio` element).

Since then, and confirmed by the testing I've just done, video block preview placeholders aren't hiding with their `video` elements. Tangentially related to this, Chrome has started to exhibit the symptoms of [SJRK-226](https://issues.fluidproject.org/browse/SJRK-226), which previously only applied to Firefox. With this in mind, I have added this model listener for **all 3** media block types (`audio`, `image`, and `video`), and have also solved SJRK-226 by making the image block preview placeholder its own element (the same one shared by the other two types).

Furthermore, because the `imageBlockEditor` grade is currently not derived from the same base as the `audioBlockEditor` and `videoBlockEditor` grades (which are both instances of `mediaBlockEditor`), this solution required a bit of duplication. The work to combine these three block type editors into `mediaBlockEditor` is described in [SJRK-175](https://issues.fluidproject.org/browse/SJRK-175), which will hopefully get some attention in the future, but may be out of scope for this issue.

Please see the commit message for this commit for more details: https://github.com/fluid-project/sjrk-story-telling/pull/77/commits/c539dd0f0b3b9554b6e72affdd770d87ab3301b9

jobara commented on src/ui/blockUi-editor.js:
```diff
@@ -114,6 +118,145 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+            uploadButtonDisabled: false
+        },
+        modelListeners: {
+            previewVisible: {
+                this: "{that}.dom.mediaPreview",
+                method: "toggle",
+                args: ["{change}.value"],
+                namespace: "previewVisibleChange"
+            },
+            progressAreaVisible: {
+                this: "{that}.dom.progressArea",
+                method: "toggle",
+                args: ["{change}.value"],
+                namespace: "progressAreaVisibleChange"
+            },
+            responseAreaVisible: {
+                this: "{that}.dom.responseArea",
+                method: "toggle",
+                args: ["{change}.value"],
+                namespace: "responseAreaVisibleChange"
+            },
+            uploadButtonDisabled: {
+                this: "{that}.dom.uploadButton",
+                method: "prop",
+                args: ["disabled", "{change}.value"],
+                namespace: "uploadButtonDisabledChange"
+            }
```
May be able to collapse this down into a single listener handler that listens to model changes on all of the specified paths and performs the necessary toggling based off the model change.

see: https://docs.fluidproject.org/infusion/development/ChangeApplierAPI.html#the-path-entry-in-a-model-listener-declaration

BlueSlug replied:
As we discussed earlier today, this would require a model listener that listens to multiple paths (which is possible) but which also needs to know which DOM element it is going to be acting on.

Short of making the names of the model values consistently refer to the selector names and automatically extracting the latter from the former (which strikes me as a brittle and overcomplicated solution), I'm not sure there's a nice way to combine these into a single listener.

jobara replied:
Let's not worry about this for the current PR and we can revisit if needed later.

cindyli commented on src/ui/blockUi-editor.js:
```diff
@@ -114,6 +118,145 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                                    },
+                                    "errorReceived": {
+                                        outputValue: {
+                                            previewVisible: true,
```
Shall the preview be visible when the upload receives an error?

BlueSlug replied:
The designs for [SJRK-396](https://issues.fluidproject.org/browse/SJRK-396) seem to suggest that the preview will be hidden when the upload receives an error. I will set the value to `false` for now, and we can review the question when SJRK-396 is picked up.

cindyli commented on src/ui/blockUi-editor.js:
```diff
@@ -114,6 +118,145 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                            func: "{that}.events.onFileSelectionRequested.fire",
+                            namespace: "fireSelectionForFileUpload"
+                        },
+                        "onUploadComplete": {
```
onUploadComplete -> onUploadSuccess? To be consistent with "onUploadError".

jobara commented on src/ui/blockUi-editor.js:
```diff
@@ -114,6 +118,145 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                    },
+                    listeners: {
+                        "{editor}.events.onMediaUploadRequested": {
+                            func: "{that}.events.onFileSelectionRequested.fire",
```
I don't think you need to add the `fire` segment at the end in the listener block.

cindyli commented on src/ui/story.js:
```diff
@@ -17,6 +17,7 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
     fluid.defaults("sjrk.storyTelling.story", {
         gradeNames: ["fluid.modelComponent"],
         model: {
+            id: null, // to be filled in on initialization
```
It would be useful to document the db structure and fields including descriptions.

jobara commented on src/ui/blockUi-timeBased.js:
```diff
@@ -28,51 +28,51 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                func: "{that}.updateMediaPreview",
                 args: ["{that}.block.model.mediaUrl"]
             },
             "onTemplateReady.bindOnMediaReady": {
-                "this": "{that}.dom.mediaPlayer.0",
+                "this": "{that}.dom.mediaPreview.0",
                 method: "addEventListener",
                 args: ["canplay", "{that}.events.onMediaReady.fire"]
             },
             "onTemplateReady.bindOnMediaDurationChange": {
-                "this": "{that}.dom.mediaPlayer.0",
+                "this": "{that}.dom.mediaPreview.0",
                 method: "addEventListener",
                 args: ["durationchange", "{that}.events.onMediaDurationChange.fire"]
             },
             "onTemplateReady.bindOnMediaEnded": {
-                "this": "{that}.dom.mediaPlayer.0",
+                "this": "{that}.dom.mediaPreview.0",
                 method: "addEventListener",
                 args: ["ended", "{that}.events.onMediaEnded.fire"]
             },
             "onTemplateReady.bindOnMediaPlay": {
-                "this": "{that}.dom.mediaPlayer.0",
+                "this": "{that}.dom.mediaPreview.0",
                 method: "addEventListener",
                 args: ["play", "{that}.events.onMediaPlay.fire"]
             },
```
This should probably be collapsed into a single function to relay the necessary events.

jobara commented on src/ui/story.js:
```diff
@@ -7,7 +7,7 @@ You may obtain a copy of the BSD License at
 */
 
-/* global fluid */
+/* global fluid, sjrk */
```
You can add "sjrk" to the eslint config so that you don't need to specify it as a global in all the files.

BlueSlug replied:
I began to work on this, but then realized it affects nearly every file (as `fluid` is [also specified](https://github.com/fluid-project/eslint-config-fluid/blob/041cd0f476fc02d0293914962c14912750e753f5/.eslintrc.json#L7) in the `.eslintrc.json` for `eslint-config-fluid` itself). Due to the already cluttered nature of this PR, I decided this is out of scope and described the work under the freshly-filed [SJRK-417](https://issues.fluidproject.org/browse/SJRK-417)

cindyli commented on src/ui/story.js:
```diff
@@ -32,8 +33,37 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+            timestampLastModified: "",
+            timestampPublished: ""
+        },
+        invokers: {
```
According to the discussion with @BlueSlug, `timestampLastModified` is created for addressing [SJRK-345 issue](https://issues.fluidproject.org/browse/SJRK-345) when we need to set up a rule to check staled stories. Since we haven't set up any rules in relation to this timestamp, it might be good to remove this code before it slips out of our mind. @BlueSlug will keep the commit hash to this code in SJRK-345 for the reference in the future when this code is useful.

BlueSlug replied:
I have added a comment with the commit hash on the SJRK-345 Jira

jobara commented on src/ui/uploader.js:
```diff
@@ -28,13 +28,20 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+            uploadState: "ready",
+            fileObjectUrl: null, // for the file preview
+            previousFileObjectUrl: null, // for file deletion on change/upload
+            serverUploadUrl: "/stories/",
```
I don't think this should be a model value, as it is likely not something that would change, and isn't part of the "state".

cindyli replied:
Some thoughts from the discussion with @BlueSlug and @jobara:

1. The upload url and method should be configurable;
2. These values make more sense to be in the "members" option rather than "model".

jobara commented on src/server/serverSetup.js:
```diff
@@ -164,11 +169,16 @@ fluid.defaults("sjrk.storyTelling.server.app.storyTellingHandlers", {
         },
-        saveStoryWithBinariesHandler: {
-            type: "sjrk.storyTelling.server.saveStoryWithBinariesHandler",
+        saveStoryHandler: {
+            type: "sjrk.storyTelling.server.saveStoryHandler",
             "route": "/stories/",
             "method": "post"
         },
+        saveStoryFileHandler: {
+            type: "sjrk.storyTelling.server.saveStoryFileHandler",
+            "route": "/stories/:id",
+            "method": "post"
+        },
```
Should review the end points for proper RESTful semantics. In this case POST to create a story makes sense. However uploading a file would likely be better to POST to /stories/:id/files. Updating specific files, or stories can use PUT and deleting a file or story can use DELETE.

jobara replied:
Would also be good to have all of the story related files, that aren't in the database, stored in a directory named off the story id. This would make organization easier as well as deleting the story.

BlueSlug replied:
As we discussed, the work to make the server API more closely follow RESTful semantics is out of scope for this Jira, so I've filed [SJRK-416](https://issues.fluidproject.org/browse/SJRK-416) to tackle this work.

The work to save files in directories named after the storyID rather than in the same directory with a filename based on the storyID still makes sense within the scope of SJRK-289.

jobara commented on src/ui/uploader.js:
```diff
@@ -28,13 +28,20 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+            // "ready" (the initial state), "uploading", "errorReceived"
+            uploadState: "ready",
+            fileObjectUrl: null, // for the file preview
+            previousFileObjectUrl: null, // for file deletion on change/upload
```
We can probably remove the need for the `previosFileObjectUrl` model path. Instead we can add a model listener for changes to the fileObjectURL and trigger a delete when it changes. Or there is a request to re-upload actually fire a delete immediately for the current fileObjectUrl if one exists. We would need to update the middleware that handles the upload. This will also make things more RESTful.

BlueSlug replied:
Would it be reasonable to count this as part of the work for [SJRK-416](https://issues.fluidproject.org/browse/SJRK-416)? If so, I can adjust the description of that Jira to explicitly include the file upload endpoints.

jobara replied:
@BlueSlug is that because of the DELETE operation?

BlueSlug replied:
That's right: since making this change will require changing the server endpoints, it might make sense to do it at that same time

BlueSlug committed:
SJRK-289: disable previewing when a file is uploading or error received

BlueSlug committed:
Merge remote-tracking branch 'upstream/master' into SJRK-289

BlueSlug committed:
SJRK-289: fix bug: keep text block from trying to reset uploader state

BlueSlug committed:
SJRK-289: rename removePreviousStoryFile middleware to be more general

`deleteStoryFile` was also considered, but the implementing function 
does not actually refer to a story in any way, so this is not accurate.

BlueSlug committed:
SJRK-289: remove superfluous intermediary JSON parsing variables

BlueSlug committed:
SJRK-289: make timeBased blockUi event listener registration more DRY

BlueSlug committed:
SJRK-289: add missing JS file ref to video block manual test page

BlueSlug committed:
SJRK-289: Update src/ui/base-page-storyEdit.js

Co-authored-by: Cindy Qi Li <cindyli@users.noreply.github.com>

BlueSlug committed:
SJRK-289: remove `timestampLastModified`, related functions and tests

BlueSlug committed:
SJRK-289: fix bug introduced in 77c8ed5, correct test count (timestamp)

The moveBlock disabled states were backfeeding into the block order, 
causing a model relay loop. How this wasn't caught by the tests is 
beyond me!

BlueSlug committed:
SJRK-289: address review items

This commit:
- adds logging to the `deleteFile` middleware
- adds descriptions for story model values
- renames `saveNewStoryToServer()` and `saveStoryToServer()` and 
associated events to be less ambiguous and more descriptive
- renames the singleFileUploader grade's `onUploadComplete` event to 
`onUploadSuccess` to better match the corresponding error event
- moves `serverUploadUrl` out of the singleFileUploader grade's model, 
instead setting it as an option

BlueSlug committed:
SJRK-289: fix bugs with file upload & related error messages

BlueSlug committed:
SJRK-289: refactor uploader to make file save route & verb configurable

Also makes some changeApplier calls easier to read and 
reorganizes/clarifies model values for blockUi.editor grade

BlueSlug committed:
Merge remote-tracking branch 'upstream/main' into SJRK-289

BlueSlug committed:
SJRK-418: solve bug by using defaultInput instead of defaultInputPath

BlueSlug committed:
SJRK-289: show/hide placeholder for all media block types, match preview

This work makes the preview placeholder image (the little icon that 
shows when there is nothing to preview) show and hide along with the 
media preview element itself. Previously, it was just the audio block 
that did this. This work also fixes SJRK-226 by making the image preview 
placeholder its own element rather than an img tag with no/empty `src` 
attribute.

The solution also includes some duplication, though SJRK-175 outlines 
the work for dealing with this in the future.

BlueSlug commented:
This PR is now ready for more review, and I've taken it out of "draft" mode. There are still some tests to add to address the deficit in test coverage that all the new code has introduced, but I doubt we'll have time to handle all of it, according to our schedule.

jobara commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
     "name": "sjrk-story-telling",
-    "version": "0.3.0",
+    "version": "1.0.0",
```
I believe we've decided this will be an 0.4.0 release instead

jobara commented on src/server/middleware/deleteFile.js:
```diff
@@ -0,0 +1,58 @@
+                throw "The specified file did not exist, and could not be deleted: " + fileToDelete;
+            }
+
+            fluid.log(fluid.logLevel.WARN, "Successfully deleted file: ", fileToDelete);
```
Seems like a log level of Info or Important would be sufficient here. 
https://docs.fluidproject.org/infusion/development/coreapi#fluidloglevel

BlueSlug replied:
Info seems like the most reasonable choice, after all, as this would fall into the category of "corresponding to verbose debugging messages", especially considering this will be logged any time an author uploads a new file for a given block

jobara commented on src/server/middleware/deleteFile.js:
```diff
@@ -0,0 +1,58 @@
+    if (request.req.body.previousFileUrl) {
+        var fileToDelete = path.join(uploadedFilesDirectory, path.basename(request.req.body.previousFileUrl));
+
+        fluid.log(fluid.logLevel.WARN, "Deleting file: ", fileToDelete, "...");
```
Seems like a log level of Info or Important would be sufficient here. 
https://docs.fluidproject.org/infusion/development/coreapi#fluidloglevel

jobara commented on src/server/requestHandlers.js:
```diff
@@ -111,50 +109,40 @@ fluid.defaults("sjrk.storyTelling.server.getStoryHandler", {
     var promise = dataSource.get({directStoryId: id});
 
-    promise.then(function (response) {
+    var noAccessErrorMessage = "An error occurred while retrieving the requested story";
```
We may need to review error messages and make them more configurable on the server. Perhaps a JIRA for the future.

BlueSlug replied:
I agree, and they should be localizable. This goes for the logging and messages in the rest of the request handlers and middleware, as well. I have filed [SJRK-419](https://issues.fluidproject.org/browse/SJRK-419) to tackle this work.

jobara commented on src/server/requestHandlers.js:
```diff
@@ -111,50 +109,40 @@ fluid.defaults("sjrk.storyTelling.server.getStoryHandler", {
+        if (response.published) {
+            request.events.onSuccess.fire(JSON.stringify(response));
+        } else {
+            fluid.log("Unauthorized: cannot access an unpublished story: " + id);
```
Maybe this one should have a Warn level log.

jobara commented on src/server/requestHandlers.js:
```diff
@@ -111,50 +109,40 @@ fluid.defaults("sjrk.storyTelling.server.getStoryHandler", {
+        }
     }, function (error) {
-        var errorAsJSON = JSON.stringify(error);
+        fluid.log("Error getting story with ID " + id + ", error detail: " + JSON.stringify(error));
```
This seems also like it should have a log level of Warn

BlueSlug replied:
Warn seems reasonable here, though if we get a large number of attempts to access stories without authorization, then we may want to drop this to "Important" if the logs are getting too big

BlueSlug replied:
We should probably sort out log backups first ;P

jobara commented on src/server/requestHandlers.js:
```diff
@@ -222,42 +188,85 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
+
+    // verify that the story exists and isn't published before continuing
+    dataSource.get({directStoryId: id}).then(function (story) {
+        if (story && !story.published && request.req.file) {
```
Is it possible for the `dataSource.get` to resolve without a story?

BlueSlug replied:
I believe this won't ever be the case, so the check for `story` is probably overkill. Removing it for now. If we discover a case where it *is* possible, then we'll want to add the same check to `sjrk.storyTelling.server.handleGetStory`

jobara commented on src/server/requestHandlers.js:
```diff
@@ -222,42 +188,85 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
+            });
         }
+    }, function (err) {
+        request.events.onError.fire({
```
Are we returning proper [http response codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) with all of the errors?

BlueSlug replied:
We aren't, currently. I've been following [the example set by `fluid.promise`](https://docs.fluidproject.org/infusion/development/promisesapi#promiserejecterror) and returning `isError: true` and a `message` string with every error, but HTTP status codes would be good to add as well. Would you like this as part of this work, or should we file a Jira to tackle it project-wide?

jobara replied:
Let's file this as a new JIRA and review things as a whole.

jobara commented on src/ui/base-page-storyEdit.js:
```diff
@@ -274,50 +364,73 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
+     * Updates the storyEditor's story and blockManager with saved story data
      *
-     * @param {Component[]} blocks - a collection of story blocks (sjrk.storyTelling.block)
+     * @param {Object} savedStoryData - story model data loaded from autosave
+     * @param {Component} story - an instance of `sjrk.storyTelling.base.page.storyEdit`
+     * @param {Component} blockManager - the storyEditor's blockManager component
+     * @param {String} sourceName - the name of the Infusion change source for this update
+     */
+    sjrk.storyTelling.base.page.storyEdit.loadStoryContent = function (savedStoryData, story, blockManager, sourceName) {
+        story.applier.change("", savedStoryData, null, sourceName);
+
+        // build the blockUIs from the story content array
+        blockManager.createBlocksFromData(savedStoryData.content);
+    };
```
It seems this whole function can be removed. Instead you could make the related invoker declaratively trigger the changeApplier. You could then have a modelListener, which could only listen to a specific source if need be, to call `blockManager.createBlocksFromData`.

BlueSlug replied:
I agree with this change and partially implemented it, but I've run into trouble setting up the model listener.

I have tried:
1. setting up the model listener on the `storyEdit` grade (top-level, where the original `loadStoryContent` invoker is)
2. moving the model listener to the `storyEditor` component (mid-level, it owns the `blockManager` and `story` components)
3. moving the model listener to the `storyEditor`'s story component (the model being listened to)
4. moving the model listener to the `blockManager` (whose invoker is called by the listener) and setting its key (i.e. path) to `"{story}.model"`
5. moving the model listener to the `blockManager` and setting its key to `"{storyEditor}.story.model"`
6. moving the model listener to the `blockManager` and setting the `segs` and `context` keys to `[]` and `"story"`, respectively

Solutions 1, 2 and 3 don't seem to be working due to timing issues. The `blockManager` component has a `createOnEvent` defined for it, and so when the `story` is being constructed, `blockManager` may not exist yet. That said, it should exist when `storyEdit` is being constructed, so this is surprising. Here is the error message in one of those cases, showing the definition of the model listener:
```
"Assertion failure - check console for more details: Error in listener record - could not resolve reference
{
    "func": "{storyEditor}.blockManager.createBlocksFromData",
    "args": ["content"],
    "includeSource": ["storyAutoload"],
    "namespace": "createEditorBlocksFromStoryData"
}
to a listener or firer. Did you miss out "events." when referring to an event firer?"
```

Solutions 3, 4 and 5 don't work either, though no errors are thrown and it fails silently.

jobara replied:
I'm having a bit of trouble understanding who should be responsible for which parts. It looks like this is a method on `sjrk.storyTelling.base.page.storyEdit` which modifies the model of the `storyEditor` subcomponent's `story` subcomponent's model. It then makes a call to the `storyEditor`'s `blockManager`'s `createBlocksFromData` method. I'm not entirely clear on the distinction between `storyEdit` and `storyEditor`. I'm guessing the former is the page, and the latter is just the portion where stories are written. Is that correct?

jobara replied:
I think I have this working with the attached changes. However, I'm still not sure of the factoring. 
[SJRK-289.patch.zip](https://github.com/fluid-project/sjrk-story-telling/files/5517903/SJRK-289.patch.zip)

BlueSlug replied:
That's correct: `storyEdit` is the page which includes the `storyEditor`, `storyViewer` and a couple other `ui` grade subcomponents

`storyEditor` is the `ui` component which has the story (which in turn has the data), a templateManager which renders everything, the binder, and the blockManager (which manages the blockUi components which each in turn have a template manager & block)

This patch worked, thanks! I've opted to use the previous event source value (`storyAutoLoad`) which was already present, rather than the new one in your path (`loadStoryContent`, named after the invoker that initiates the change). The reason for this is that the former is the source for the changes triggered in `sjrk.storyTelling.base.page.storyEdit.createNewStoryOnServer`, and we don't need/want to call `createBlocksFromData` at this point.

jobara commented on src/ui/base-page-storyEdit.js:
```diff
@@ -334,6 +447,43 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
+     * @param {String} newState - the new state of the value
+     * @param {Component} storyEdit - an instance of `sjrk.storyTelling.base.page.storyEdit`
+     */
+    sjrk.storyTelling.base.page.storyEdit.updateUploadCounters = function (previousState, newState, storyEdit) {
+        var counters = fluid.copy(storyEdit.model.uploadCounters);
+
+        if (previousState === "errorReceived") {
+            counters.errors--;
+            counters.uploads++;
+        } else if (previousState === "uploading") {
+            counters.uploads--;
+
+            if (newState === "errorReceived") {
+                counters.errors++;
+            }
+        } else if (previousState === "ready") {
+            counters.uploads++;
+        }
+
+        storyEdit.applier.change(["uploadCounters"], counters);
```
Unless this is handled in a model relay, it doesn't appear that there are any limits on the range. At least we shouldn't allow the count to go below 0.

BlueSlug replied:
I don't think it's possible for the counters to go below 0, as neither of them is decremented when the `previousState` is not one of the three cases handled by the outer `if` block. Assuming model relays & the `changeApplier` are synchronous, i.e. they won't be able to run simultaneously, then the same counter could not be decremented without a corresponding initial increment. I can add the following just before the `changeApplier` call, if you think it would be helpful:
```JavaScript
// ensure the counters are not negative
counters.uploads = Math.max(counters.uploads, 0);
counters.errors = Math.max(counters.errors, 0);
```

There may also be a way to achieve the aims of this function (and enforce its minima) more elegantly via model relays and transform functions.

jobara commented on src/ui/base-page-storyEdit.js:
```diff
@@ -384,55 +534,74 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
+     * @param {Component} storyEditPage - an instance of `sjrk.storyTelling.base.page.storyEdit`
+     * @param {String} viewPageUrl - the URL for the story View page
+     * @param {Component} story - an instance of `sjrk.storyTelling.story`
+     * @param {publishSuccessCallback} successCallback - to call on successful publish
```
I wonder if this should just return a promise instead?

BlueSlug replied:
I think, given that it's called from an invoker which is in turn called in an event listener, I'm not sure how a return value could be accessed.

To stay consistent with the rest of the project, it might make more sense to pass in an event to fire which can return a promise, and then a listener to that event can go from there.

BlueSlug replied:
I've opted to add an `onStoryPublishSuccess` event that returns the story ID, which allowed me to simplify the function a bit and use listeners to do the rest. I think it's worth revisiting the idea of how the rest of the system could be adapted if `publishStory` returns a promise, though I'm not sure at this point what that would look like

jobara commented on src/ui/blockUi-editor.js:
```diff
@@ -16,12 +16,34 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
+                source: "{block}.model.lastInOrder",
+                target: "moveBlockDownDisabled",
+                singleTransform: {
+                    type: "fluid.transforms.value"
```
should this be `fluid.transforms.identity`?

BlueSlug replied:
These two functions are aliases of one another, and so I chose to use `value` since it's possibly easier to understand at a glance. That said, an Infusion developer would be used to seeing `identity`, and we do use the latter in our tests. Do you have a preference?

jobara replied:
identity might just be clearer about the expectations of what will happen, which is to say, no model transformation.

jobara commented on src/ui/blockUi-timeBased.js:
```diff
@@ -28,51 +28,42 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
             }
         },
+        // describes the relationship of DOM events to associated Infusion events
+        mediaPlayerEventMapping : {
+            canplay: "{that}.events.onMediaReady.fire",
+            durationchange: "{that}.events.onMediaDurationChange.fire",
+            ended: "{that}.events.onMediaEnded.fire",
+            play: "{that}.events.onMediaPlay.fire"
+        },
```
Another option here would be to just have the Infusion event names on the right hand side. In that case `sjrk.storyTelling.blockUi.timeBased.registerMediaPlayerEvents` would need to take in `that` and would construct the output even firer like `that.events[eventName].fire'

jobara commented on src/ui/uploader.js:
```diff
@@ -44,98 +54,157 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
+                    uploadOptions.completionEvent.fire(data);
+                },
+                error       : function (jqXHR, textStatus, errorThrown) {
+                    fluid.log(jqXHR, textStatus, errorThrown);
```
maybe this one should be have a Warn level

jobara commented on src/ui/uploader.js:
```diff
@@ -44,98 +54,157 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
-        }
+    sjrk.storyTelling.block.singleFileUploader.updateFileObjectInformation = function (that) {
+        that.applier.change("previousFileObjectUrl", that.model.fileObjectUrl);
+        that.applier.change("fileObjectUrl", "");
```
This could be done declaratively and in a single change request.

BlueSlug committed:
SJRK-289: set version to 0.4.0 ahead of next release

We won't be calling this version "1.0.0" after all

BlueSlug committed:
SJRK-289: make misc improvements to the code to satisfy review feedback

- Changed the logging level of various calls to `fluid.log`
- removed extraneous checks
- made the `mediaPlayerEventMapping` more readable & more easily 
configured
- consolidated an invoker into a declarative changeApplier call

jobara commented on src/server/middleware/deleteFile.js:
```diff
@@ -0,0 +1,58 @@
+
+            fluid.log(fluid.logLevel.INFO, "Successfully deleted file: ", fileToDelete);
+        } catch (err) {
+            fluid.log(fluid.logLevel.FAIL, "Error deleting file \"", fileToDelete, "\".");
+            fluid.log(fluid.logLevel.FAIL, "Error detail: ", err.toString());
```
Are these really Fail level? Will the system terminate shortly after this?

jobara commented on src/ui/base-page-storyEdit.js:
```diff
@@ -274,50 +364,73 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
+        // The "story created" moment is now
+        story.applier.change("timestampCreated", new Date().toISOString());
+
+        var serverSavePromise = sjrk.storyTelling.base.page.storyEdit.updateStoryOnServer(storySaveUrl, story.model);
```
What is in the `story.model` at this point? Is it just the `timestampCreated`? If so, I wonder if it would make more sense to have the server create the time stamp. In terms of the timestamp currently what happens if the `serverSavePromise` is rejected?

jobara commented on src/ui/base-page-storyEdit.js:
```diff
@@ -334,6 +447,40 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
+
+        if (previousState === "errorReceived") {
+            counters.errors--;
+            counters.uploads++;
```
Is it not possible to go from "errorReceived" to "ready"?

BlueSlug replied:
As we discussed at our daily check-in today, this is possible when the `resetUploadState` invoker is called, which is currently only on block deletion. I'll add a check for it in this `if` block

BlueSlug committed:
SJRK-289: make the autoload work via Infusion rather than JavaSrript

Co-Authored-By: Justin Obara <obara.justin@gmail.com>

BlueSlug committed:
SJRK-289: make more improvements from code review suggestions (see list)

- set timestampCreated after the new story is created, not before
- add missing case for uploadState transition from `errorReceived` to 
`ready`
- change log level for deleteFile error messages to better reflect 
situation
- use `identity` instead of `value` for consistency and better 
readability

BlueSlug committed:
SJRK-289: remove callback & args for publishStory(), add onSuccess event

jobara commented on src/ui/base-page-storyEdit.js:
```diff
@@ -26,6 +26,17 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
+            "blockManager.loadStoryContentIntoBlockUIs": {
+                record: {
+                    "{storyEditor}.story.model": {
+                        func: "{that}.createBlocksFromData",
```
Realizing it's a bit ambiguous to use `{that}` here. I think we should update this to use `{blockManager}` or some other appropriate name that is more explicit.

BlueSlug committed:
SJRK-289: use explicit component name for less ambiguity

BlueSlug committed:
SJRK-289: add robots.txt and static file folder, hosting rules

BlueSlug committed:
SJRK-289: fix autosave bug in story creation process

The bug in question was due to the fact that the source that was 
(previously) provided in the call to `createNewStoryOnServer` was 
excluded by the `saveStoryToAutosave` model listener on the editor's 
story. The solution was to give the creation call its own source

jobara commented:
Due to the deadline for releasing these features we will need to merge without having finished writing tests, and thus will have reduced code coverage.

jobara commented:
merged at 125f245785f2b76487d168d6a39908dd05ed788a
